firewood/src/merkle/node.rs

```diff
@@ -490,6 +495,92 @@ pub enum EncodedNodeType {
     },
 }
 
+// TODO: probably can merge with `EncodedNodeType`.
+#[derive(Debug, Deserialize)]
+struct EncodedBranchNode {
+    chd: Vec<(u64, Vec<u8>)>,
+    data: Option<Vec<u8>>,
+    path: Vec<u8>,
+}
+
+// Note that the serializer passed in should always be the same type as T in EncodedNode<T>.
+impl Serialize for EncodedNode<PlainCodec> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let n = match &self.node {
+            EncodedNodeType::Leaf(n) => {
+                let data = Some(n.data.to_vec());
+                let chd: Vec<(u64, Vec<u8>)> = Default::default();
+                let path = from_nibbles(&n.path.encode(true)).collect();
+                EncodedBranchNode { chd, data, path }
+            }
+            EncodedNodeType::Branch { children, value } => {
+                let chd: Vec<(u64, Vec<u8>)> = children
+                    .iter()
+                    .enumerate()
+                    .filter_map(|(i, c)| {
+                        c.as_ref().map(|c| {
+                            if c.len() >= TRIE_HASH_LEN {
+                                (i as u64, Keccak256::digest(c).to_vec())
+                            } else {
+                                (i as u64, c.to_vec())
+                            }
+                        })
+                    })
+                    .collect();
+
+                let data = value.as_ref().map(|v| v.0.to_vec());
+                EncodedBranchNode {
+                    chd,
+                    data,
+                    path: Vec::new(),
+                }
+            }
+        };
+
+        let mut s = serializer.serialize_tuple(3)?;
+        s.serialize_element(&n.chd)?;
+        s.serialize_element(&n.data)?;
+        s.serialize_element(&n.path)?;
+        s.end()
+    }
+}
+
+impl<'de> Deserialize<'de> for EncodedNode<PlainCodec> {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let node: EncodedBranchNode = Deserialize::deserialize(deserializer)?;
+        if node.chd.is_empty() {
+            let data = if let Some(d) = node.data {
+                Data(d)
+            } else {
+                Data(Vec::new())
+            };
+
+            let path = PartialPath::from_nibbles(Nibbles::<0>::new(&node.path).into_iter()).0;
+            let node = EncodedNodeType::Leaf(LeafNode { path, data });
+            Ok(Self::new(node))
+        } else {
+            let mut children: [Option<Vec<u8>>; BranchNode::MAX_CHILDREN] = Default::default();
+            let value = node.data.map(Data).filter(|data| !data.is_empty());
```

Comment:
I think I asked this somewhere else, but how do we different empty data and no data?

Reply:
Good question, this is following existing encoding style as https://github.com/ava-labs/firewood/blob/main/firewood/src/merkle/node.rs#L590 and https://github.com/ava-labs/firewood/blob/main/firewood/src/merkle/node/branch.rs#L119. But probably I should not do that for MerkleDB encoding schema. I will remove it.

Reply:
Created an issue for fixing existing Bincode encoding schema https://github.com/ava-labs/firewood/issues/433